mongo/client.go

```diff
@@ -1030,3 +1030,12 @@ func (c *Client) createBaseCursorOptions() driver.CursorOptions {
 		ServerAPI:      c.serverAPI,
 	}
 }
+
+// SetZeroTimeout sets the Timeout to 0 for an existing client.
+//
+// Deprecated: This setter is only used for internal testing of the Timeout feature. Modifying the Timeout of an existing
+// client can result in undefined behavior.
+func (c *Client) SetZeroTimeout() {
```

Comment:
This was a tough design decision. `Client`, `Database` and `Collection` objects are usually created through the top-level [`createEntities`](https://github.com/mongodb/specifications/blob/master/source/client-side-operations-timeout/tests/override-operation-timeoutMS.yml#L15) field on unified test format files. However, as stated in the PR description, we need to set a Timeout of `0` on the caller of any operation that has a `timeoutMS` field. Modifying `timeout` after the `Client`, `Database` or `Collection` has been created is not possible without this setter since it is an unexported field. Let me know if you have a better idea of how to do this, but adding these deprecated setters was the best thing I could come up with. There are other CSOT spec tests that set `timeoutMS` on individual operations, so I think we might have to add operation-level support.

Reply:
This seems tricky. IIUC the goal is to only apply the `Timeout` logic to one operation. But there is no way to accomplish that with public API.

I prefer avoiding public API only intended for internal testing. Public API can always be added later.

IIRC we want to avoid having users add values to contexts. But maybe this could be done internally for this case. Example:

```go
// In file: internal/csot_util.go

package internal

import "context"

type timeoutKey struct{}

// MakeTimeoutContext returns a new context with Client-Side Operation Timeout (CSOT) feature-gated behavior.
// Setting a Timeout on a single operation is not supported in public API.
// This may be unnecessary after GODRIVER-2348 is resolved.
func MakeTimeoutContext(ctx context.Context) context.Context {
	return context.WithValue(ctx, timeoutKey{}, true)
}

func IsTimeoutContext(ctx context.Context) bool {
	return ctx.Value(timeoutKey{}) != nil
}
```

Reply:
Should we not be doing this through the `clientEntity` struct?  For exmaple, in`executeListDatabases` we get our client from

```
client, err := entities(ctx).client(operation.Object)
```

It seems that if we want the client to have a zero timeout for this test case, then the options used to create a new `clientEntity`  should include `Timeout =0`.  I think it would require some tweaks [here](https://github.com/mongodb/mongo-go-driver/blob/master/mongo/integration/unified/unified_spec_runner.go#L251), specifically.

We could add a new `entityType` called `"zeroTimeoutClient"` or something of that nature which takes the exact options as the existing client but includes `opts.Timeout = 0`.  Then in our branch we could do something like this:

```go
	if operation.TimeoutOnCaller {
	        zeroTimeoutCallback, err := entities(ctx).client("zeroTimeoutClient")
		client = zeroTimeoutCallback()
	}
```

Reply:
Both excellent alternatives! I agree we want to avoid the `SetZeroTimeout` deprecated public API. I've opted for a version of @kevinAlbs 's suggestion that sets `timeoutKey` on the `Context` whenever we want to opt-in to CSOT behavior (not _just_ for testing). This solution requires minimal code changes, and we'll be able to remove the `internal.IsTimeoutContext` checks with GODRIVER-2348 in version 2.0. I know passing values in contexts is usually a [faux pas](https://pkg.go.dev/context#:~:text=Use%20context%20Values%20only%20for,simultaneous%20use%20by%20multiple%20goroutines.), but we do something similar for sessions since they need to be set with public API and accessed within the `x` package. 

@prestonvasquez your solution also seems viable, and you are correct that "registering" one extra client would not be a huge performance hit in our testing. Let me know what you think of the context solution, as I believe we can avoid the numerous `TimeoutOnCaller` checks and the extra client entities.